Add a flip-filter

mattiasb opened the pull request:
##### SUMMARY

Add a flip-filter that can be used to apply arguments in reverse order to another filter.

Useful in combination with `map` and similar filters.

Also, see: #46255 and #46215 .

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

No idea.

##### ANSIBLE VERSION
```paste below
ansible --version
ansible 2.8.0.dev0 (feature/flip-filter 97ef190cd6) last updated 2018/10/01 13:18:12 (GMT +200)
  config file = /home/mattiasb/.ansible.cfg
  configured module search path = [u'/home/mattiasb/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mattiasb/Code/github.com/ansible/ansible/lib/ansible
  executable location = /home/mattiasb/Code/github.com/ansible/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

##### ADDITIONAL INFORMATION

From the documentation added for this PR:

--------------------------

You can use the `flip` filter to apply a filter but with arguments flipped in
reverse order. Example ::

    {{ list1 | flip('difference', list2) }}

… is the same as:

    {{ list2 | difference(list1) }}

This becomes useful with filters like `map` where you can't easily re-order the
arguments yourself.

Say that you have a role that takes a list of users to add to a system. You want
to ensure some default values and apply some overrides before using the value in
your role's task. You could do it like this:

`roles/users/defaults/main.yml`:
```yaml
    # A list of user objects
    users: []
```

`roles/users/vars/main.yml`:
```yaml
    # Ensure we don't delete users and that we never overwrite passwords after
    # creation
    _users_overrides:
      state: present
      update_password: on_create

    # Add users to `guest` by default and generate an ssh-key.
    _users_defaults:
      groups:
        - guest
      generate_ssh_key: true

    # A list of users with overrides and defaults applied (this is what your
    # tasks will use)
    _users: >-
      {{ users
       | map('flip', 'combine', _users_defaults)
       | map('combine', _users_overrides)
       | list
      }}
```

Ping @bcoca and @gundalow who's been involved on my journey to get my use-case solved. ☺

mattiasb committed:
Add flip filter

Add a flip-filter that can be used to apply arguments in reverse order to
another filter.

Inspired by the `flip` function in functional programming languages.

mattiasb committed:
Filter tests: Test combine-filter

The combine-filter was untested so add the examples from the filter
documentation[1] and make it a test.

1: https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#combining-hashes-dictionaries

mattiasb commented:
Open question: Should I try to push for this upstream instead?

bcoca commented:
possibly, seems like a good thing to have for templates in general

I've been thinking of that lately specially since we started having conflicts between our filters/tests and jinaj2's new ones

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/86282/53/tests

mattiasb commented:
> CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/86282/53/tests

Updated the kwargs-test to use the built-in `combine` filter again.

Also rebased against devel.

mattiasb commented:
Added [a PR](https://github.com/pallets/jinja/pull/906) upstream now. Let's see what they say.

mattiasb commented:
CI is failing with:
```
fatal: [testhost]: FAILED! => {
05:56     "msg": "The conditional check ''one-arg-identity' | upper == 'one-arg-identity' | flip('upper')' failed. The error was: 'AnsibleEnvironment' object has no attribute 'call_filter'"
05:56 }
```
Is this running against some old version of jinja or something? I can't reproduce this on my side.

mattiasb commented:
One issue I've found today is that this won't work with `select` since it takes a test as first argument and `flip` is a filter.

/me feeling frustrated

mattclay commented:
@mattiasb Yes, on our centos6 test container we're using version 2.2.1.

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/87520/46/tests

bcoca commented:
@mattiasb are you going to have time to fix the test issue? It needs to happen soon to make it into 2.8

mattiasb commented:
@bcoca no sorry. If I remember correctly I came to the conclusion that you really can't implement something like `flip` with Jinja 2.2.1.

Akasurde commented:
Hi @mattiasb, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
